tests/test_highlighting/test_highlighting_adventure_en_03.py

```diff
@@ -0,0 +1,276 @@
+from tests.Highlighter import HighlightTester
+
+class HighlighterTestLeveL3(HighlightTester):
+
+
+    def test_1(self):
+        self.assert_highlighted_chr_multi_line(
+            "print hello world!",
+            "KKKKK TTTTTTTTTTTT",
+            level="level3",lang='en')
+
+
+    def test_2(self):
+        self.assert_highlighted_chr_multi_line(
+            "animals is dog, cat, kangaroo",
+            "TTTTTTT KK TTT  TTT  TTTTTTTT",
+            "print animals at random",
+            "KKKKK TTTTTTT KK KKKKKK",
+            level="level3",lang='en')
+
+
+    def test_3(self):
+        self.assert_highlighted_chr_multi_line(
+            "animals is dog, cat, kangaroo",
+            "TTTTTTT KK TTT  TTT  TTTTTTTT",
+            "add penguin to animals",
+            "KKK TTTTTTT KK TTTTTTT",
+            "remove cat from animals",
+            "KKKKKK TTT KKKK TTTTTTT",
+            "print animals at random",
+            "KKKKK TTTTTTT KK KKKKKK",
+            level="level3",lang='en')
+
+
+    def test_4(self):
+        self.assert_highlighted_chr_multi_line(
+            "print Your story",
+            "KKKKK TTTTTTTTTT",
+            level="level3",lang='en')
+
+
+    def test_5(self):
+        self.assert_highlighted_chr_multi_line(
+            "animals is 🦔, 🐿, 🦉, 🦇",
+            "TTTTTTT KK T  T  T  T",
+            "print He now hears the sound of an animals at random",
+            "KKKKK TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT KK KKKKKK",
+            level="level3",lang='en')
+
+
+    def test_6(self):
+        self.assert_highlighted_chr_multi_line(
+            "print He hears a sound",
+            "KKKKK TTTTTTTTTTTTTTTT",
+            "animals is 🐿, 🦔, 🦇, 🦉",
+            "TTTTTTT KK T  T  T  T",
+            "animal is ask What do you think it is?",
+            "TTTTTT KK KKK TTTTTTTTTTTTTTTTTTTTTTTT",
+            "add animal to animals",
+            "KKK TTTTTT KK TTTTTTT",
+            "print it was a animals at random",
+            "KKKKK TTTTTTTTTTTTTTTT KK KKKKKK",
+            level="level3",lang='en')
```

Comment:
I agree with @rix0rrr that there are many many tests. I appreciate that you want to cover a lot, but it should also still be readable. 

Can we make it fewer tests in this way by limiting the test set to one line tests for concepts in each level (print, ask, echo, forward, turn in level 1 etc.)

Add a few multiline ones. We don't need all those multiline test, if we just have a few that is ok. We could even write a little script to generate multiline ones from simpler ones.

Reply:
I tested a new solution, it splits the check in 2 different assert (In and Equal) but it can be enough

Reply:
> These tests allow to cover all the codes displayed on Hedy. This seemed important to me, because it ensures that, no matter what changes are made to the syntax coloring, the codes displayed will always be valid. There is necessarily some redundancy between them, but it allows to have a base of functionality to cover. 

"Some" redundancy is definitely okay. But For example, it would be fine to have 2-3 tests that cover (different aspects!) of the `print` command. But there are degrees here. You easily have 100s of tests here covering the print statement! That seems a bit much.

Can we at least agree that the following principle is a good one to strive for?

**We are looking for the *minimum* amount of tests that covers 100% of the functionality**

If we agree that's the goal, then now of course the question is how do we find that optimum point? 

* What's the minimum amount of tests? Could we remove one? I'm not sure.
* Are we sure we've covered 100% of the functionality? Might there be a bug we're not covering? I'm not sure.

There are no easy mathematical answers here. We cannot *compute* this number. But we can approximate it based on intuition, and approach it over time as we refine our test suite.

Since it's easy to add tests when their need is demonstrated, but hard to know whether a test can be removed because it's not pulling its weight... I would err on the side of conservatism. 2-3 tests for every feature ("behavior"), and let the rest sort itself out over time as we discover bugs.

However you decide to approach it, my intuition says the number you have now is too high.

I appreciate the thoroughness, and you're no doubt approaching this with a mathematical mindset... but you have to consider that software engineering is "people work" as much as it is computer work. Code is never *done*, it will always be updated, by people, and you have to allow for that future change.